<center><h1><strong><font color="blue"> Advanced Programming for Data Science (APDS)</font></strong></h1></center>
<center><h2><strong><font color="black"> Week II · Definite Iteration: `for` Loops in Python</font></strong></h2></center>

<center><img alt="" src="images/apds-2026-02-definite-looping-FOR-header.png" /></center>

<center><h3><strong><font color="blue"> © Taufik Sutanto · UIII · Odd Semester 2026</font></strong></h3> </center>

In this notebook, we will move from a simple summation to practical operations on
collections and matrices. I encourage you to predict each result before running
the code, then explain any difference between your prediction and the output.

### Learning goals

By the end of this notebook, you should be able to:

- Explain how a `for` loop obtains successive items from an iterable.
- Use `range()` with appropriate start, stop, and step values.
- Trace a loop and construct accumulators and transformed lists.
- Use direct iteration, `enumerate()`, and `zip()` appropriately.
- Interpret sequential and nested loops, including dependent inner bounds.
- Apply `break`, `continue`, and loop `else` in simple examples.
- Diagnose common mistakes and check results against small, known cases.

### Lesson outline

1. [From summation to iteration](#summation)
2. [`range()` and basic loop patterns](#ranges)
3. [Iteration over collections](#collections)
4. [Data transformations and memory](#transformations)
5. [Sequential and nested loops](#nested)
6. [Matrices and flattening](#matrices)
7. [Nested loops with dependent bounds](#dependent)
8. [Loop control: `break`, `continue`, and `else`](#control)
9. [Common mistakes and an integrated example](#practice-checks)
10. [Exercises](#exercises)
11. [Review and next steps](#review)

Additional resources: [Worked solutions](#solutions) · [References](#references)

### Review of the previous lesson

Before we begin, briefly revisit the following questions and topics:

1. Why use Python?
2. How do interpretation and compilation relate to program execution?
3. Where is Python used: web applications, mobile-related development, AI, and big data?
4. What are the essential rules of Python syntax?
5. Which basic scalar types have we encountered?
6. How do we use Jupyter Notebook and Markdown?
7. What are Python's limitations, and how might they affect our choice of tools?

<a id="summation"></a>
## 1. From summation to iteration
### A problem associated with Gauss

A familiar classroom story associates the young Carl Friedrich Gauss with the
rapid summation of the integers from 1 to 100. We will use the mathematical
problem as a teaching example, without relying on the historical details.

How would you calculate the answer? Pairing the first and last terms gives
$1 + 100 = 101$, $2 + 99 = 101$, and so on. Fifty such pairs give $5050$.
More generally, for a positive integer $n$,

$$1 + 2 + \cdots + n = \frac{n(n+1)}{2}.$$

Let us first solve the problem one term at a time. This less compact approach
makes the role of a loop especially clear.

### Summing the integers from 1 to 100

We need an **accumulator**: a variable that stores a running result. We will call
it `total`, leaving Python's built-in function `sum()` available for later use.

**Input:** a positive integer $n = 100$.  
**Output:** the sum of the integers from 1 through $n$.

```text
total ← 0
FOR i ← 1 TO n, inclusive
    total ← total + i
END FOR
REPORT total
```

Initialization belongs before the loop. If we reset `total` to zero during every
iteration, earlier contributions would be lost.

<center><img alt="" src="images/3_Looping_Components.png" style="width: 800px; height: 424px;" />

### Components of a Python `for` loop

```python
total = 0
for i in range(1, 101):
    total += i
print(total)
```

| Component | Example | Role |
|---|---|---|
| Initialization | `total = 0` | Establish the starting state. |
| Loop target | `i` | Receive the next item on each iteration. |
| Iterable | `range(1, 101)` | Supply the items to visit. |
| Header | `for ... in ...:` | Introduce the loop and end with a colon. |
| Indented body | `total += i` | Perform the repeated work. |
| Statement after the loop | `print(total)` | Run after iteration finishes. |

Use consistent indentation, conventionally four spaces per level. Python uses
indentation to identify the body; it has no `END FOR` statement.

### What does “definite iteration” mean?

A loop repeats a block of statements. In this lesson, **definite iteration** means
traversing a specified finite collection or range. A Python `for` loop requests
items from an **iterable**, an object that can supply items successively.

- The loop target is a value, not necessarily an integer index.
- The body may use that value or simply repeat an action.
- `range()` has a default step of 1, but this is not a rule for every iterable.
- A negative step is valid when counting downward is appropriate.
- Range boundaries may be calculated from variables; they need not be literals.
- Assigning to the loop target is legal, but it does not change which item the
  iterator supplies next. We will demonstrate this shortly.

A `while` loop instead repeats while a condition remains true. We focus here on
`for` loops over finite data. More generally, Python also permits iterables whose
length is unknown in advance, including infinite ones, so `for` does not itself
guarantee a predetermined iteration count or termination.

<a id="ranges"></a>
## 2. `range()` and basic loop patterns

Python's `for` syntax is close to the pseudocode we have just used. The main
difference to watch is the treatment of the final boundary in `range()`.

### Understanding `range()`

| Form | Meaning |
|---|---|
| `range(stop)` | Start at 0; use step 1. |
| `range(start, stop)` | Start at `start`; use step 1. |
| `range(start, stop, step)` | Use the specified starting value and step. |

**The stop value is excluded.** For a positive step, values remain below `stop`;
for a negative step, they remain above it. In these examples, all arguments are
integers and the step must be nonzero.

A `range` is a compact sequence description, rather than a list containing every
integer. We use `list(...)` below only to make its values visible. See the official
[range documentation](https://docs.python.org/3/library/stdtypes.html#ranges).

In [1]:
# The default starting value is 0; the stop value 7 is excluded.
list(range(7))

[0, 1, 2, 3, 4, 5, 6]

In [2]:
# An explicit starting value changes the first item, not the stop rule.
list(range(1, 7))

[1, 2, 3, 4, 5, 6]

### Implementing the summation

To include 100, use a stop value of 101. The expected result is **5050**.

In [3]:
# In Python, the # symbol begins a comment.
total = 0
for i in range(1, 101):
    total = total + i  # This can also be written as total += i.
print(total)

5050


### A shorter update expression

For the numeric values used here, `x += k` performs the same update as
`x = x + k`. This is convenient for accumulators.

In [4]:
x = 3
x = x + 3
print(x)

x = 3
x += 3
print(x)

6
6


### Trace a small example before trusting a large one

After processing integer $i$, the accumulator should equal $1 + 2 + \cdots + i$.
This statement describes a **loop invariant**: a property maintained as the loop
progresses. For the first five integers, the running totals are
$1, 3, 6, 10, 15$.

In [5]:
running_total = 0
print(f"{'i':>2} {'before':>7} {'after':>7}")
for i in range(1, 6):
    before = running_total
    running_total += i
    print(f"{i:2d} {before:7d} {running_total:7d}")

n = 100
formula_total = n * (n + 1) // 2  # // gives integer division here.
print("Formula:", formula_total)
print("Built-in sum:", sum(range(1, n + 1)))

 i  before   after
 1       0       1
 2       1       3
 3       3       6
 4       6      10
 5      10      15
Formula: 5050
Built-in sum: 5050


### Repetition without using the loop target

Sometimes the repeated action is identical each time. The name `_` conventionally
indicates that we do not need the current value.

In [6]:
for _ in range(3):
    print("Hello")

Hello
Hello
Hello


### A first example of direct iteration

A loop can visit list elements directly, even when their types differ. In this
example, `item` receives each element rather than its position.

In [7]:
items = [2, 'a', True, (2, 3)]
for item in items:
    print(item)

2
a
True
(2, 3)


### Steps, inclusive intentions, and empty ranges

Predict the difference between the two examples below: one excludes 100, while
the other includes it. Writing down the intended endpoints first helps prevent
an **off-by-one error**.

In [8]:
# Even integers from 2 up to, but excluding, 100.
print(list(range(2, 100, 2)))

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98]


In [9]:
# Even integers from 2 through 100, inclusive.
for number in range(2, 101, 2):
    print(number, end=' ')
print()  # End the output line.

2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94 96 98 100 


In [10]:
# A negative step is appropriate for a decreasing sequence.
for number in range(30, 1, -5):
    print(number)

30
25
20
15
10
5


In [11]:
print("Countdown:", list(range(5, -1, -1)))
print("Equal endpoints:", list(range(3, 3)))
print("Wrong direction:", list(range(1, 5, -1)))

empty_total = 0
for number in range(0):
    empty_total += number
print("Accumulator after an empty loop:", empty_total)

Countdown: [5, 4, 3, 2, 1, 0]
Equal endpoints: []
Wrong direction: []
Accumulator after an empty loop: 0


### Variable boundaries are valid

`range(start, stop)` is evaluated when execution reaches the loop header. Changing
a variable used to construct that range does not alter the range already created.
Similarly, rebinding the loop target does not advance or rewind its iterator.

In [12]:
start = 1
stop = 101
total = 0
for i in range(start, stop):
    total += i
print(total)

5050


In [13]:
stop = 4
supplied_values = []
for i in range(stop):
    supplied_values.append(i)
    stop = 1  # Does not change the existing range(4).
    i = 99    # Rebinds i now; the next iteration supplies the next range item.

print("Values supplied by the range:", supplied_values)
print("Final values of stop and i:", stop, i)

Values supplied by the range: [0, 1, 2, 3]
Final values of stop and i: 1 99


<a id="collections"></a>
## 3. Iteration over collections

Prefer direct iteration when you need values. Use `enumerate()` when you need
both a position and a value, and `zip()` when you need corresponding items from
multiple iterables. Tuple unpacking, as in `position, name`, assigns the two
parts of each pair to separate variables.

In [14]:
print("Characters:")
for character in "DATA":
    print(character, end=' ')
print()

print("Numbered participants:")
names = ["Alya", "Bima", "Citra"]
for position, name in enumerate(names, start=1):
    print(position, name)

print("Dictionary entries:")
scores = {"Alya": 80, "Bima": 75, "Citra": 90}
for name, score in scores.items():
    print(name, score)

Characters:
D A T A 
Numbered participants:
1 Alya
2 Bima
3 Citra
Dictionary entries:
Alya 80
Bima 75
Citra 90


`enumerate(..., start=1)` supplies convenient display numbers; it does not change
the list's zero-based indexing. Iterating directly over a dictionary supplies its
keys; `.items()` supplies key–value pairs.

`zip()` stops when its shortest input is exhausted. When equal lengths are part
of the intended data contract, `zip(..., strict=True)` (Python 3.10+) raises an
error if one input ends earlier. See the official
[built-in function descriptions](https://docs.python.org/3/library/functions.html).

In [15]:
names = ["Alya", "Bima", "Citra"]
marks = [80, 75, 90]
for name, mark in zip(names, marks, strict=True):
    print(f"{name}: {mark}")

print("Default zip with unequal lengths:", list(zip(names, [80, 75])))

Alya: 80
Bima: 75
Citra: 90
Default zip with unequal lengths: [('Alya', 80), ('Bima', 75)]


<a id="transformations"></a>
## 4. Data transformations and memory

In data science, we often apply the same transformation to many observations.
Transformations such as square roots, logarithms, and reciprocals can change a
distribution's shape. Their suitability depends on the data and the analytical
objective; none is a universal remedy for skewness.

We will use the **reciprocal transformation**, $y = 1/x$, to practise iteration.
The examples contain positive, nonzero values. A reciprocal is undefined at
zero and reverses the order of positive values: larger inputs produce smaller
outputs. Other transformations have their own domain restrictions.

<center><img alt="" src="images/3_data_transformation.png" style="width: 769px; height: 404px;" /></center>

### Method 1: Build a new list

Start with an empty list and append each transformed value. `append()` adds an
element to the existing list. This approach retains the original data. To keep
the output compact, `[:5]` displays only the first five transformed values;
every element is still transformed.

In [16]:
data = [99, 89, 80, 50, 25, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
reciprocals = []
for value in data:
    reciprocals.append(1 / value)

print("Original:", data)
print("First five reciprocals:", reciprocals[:5])

Original: [99, 89, 80, 50, 25, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
First five reciprocals: [0.010101010101010102, 0.011235955056179775, 0.0125, 0.02, 0.04]


### Method 2: Replace elements in place

A new output list needs storage that grows with the number of observations.
This is often worthwhile because it preserves the input. It does **not** mean
that `append()` copies the entire list on every call.

When the original values are no longer needed, we can replace the elements of
the existing list. This avoids a second full list container, although newly
computed numeric objects still require memory. Any other variable referring to
the same list will also observe the replacements.

In [17]:
data = [99, 89, 80, 50, 25, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
for index in range(len(data)):
    data[index] = 1 / data[index]
print("First five reciprocals:", data[:5])

First five reciprocals: [0.010101010101010102, 0.011235955056179775, 0.0125, 0.02, 0.04]


### Method 3: Use `enumerate()` for positions and values

`enumerate()` makes both pieces of information available. Replacing elements
while keeping the list's length fixed is suitable for this example. Adding or
removing elements during traversal is a different operation and can cause items
to be skipped or revisited.

In [18]:
data = [99, 89, 80, 50, 25, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
for index, value in enumerate(data):
    data[index] = 1 / value
print("First five reciprocals:", data[:5])

First five reciprocals: [0.010101010101010102, 0.011235955056179775, 0.0125, 0.02, 0.04]


### Method 4: List comprehensions

A list comprehension expresses a simple list-building operation compactly:

```python
new_list = [expression for item in iterable]
```

It creates a new list. I recommend using it when the transformation remains easy
to read; an explicit loop is clearer for several steps or complicated control
flow. Prefer clarity over assuming that one form is always faster. The
[Python tutorial on comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)
provides additional examples.

In [19]:
data = [99, 89, 80, 50, 25, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
reciprocals = [1 / value for value in data]
print([round(x, 4) for x in reciprocals])

[0.0101, 0.0112, 0.0125, 0.02, 0.04, 0.0833, 0.0909, 0.1, 0.1111, 0.125, 0.1429, 0.1667, 0.2, 0.25, 0.3333, 0.5, 1.0]


### Choosing an approach

| Approach | Preserves the input list? | Creates another full list? | Useful when |
|---|---|---|---|
| Loop with `append()` | Yes | Yes | Each item requires several operations. |
| Index-based replacement | No | No | Positions are central to the task. |
| Replacement with `enumerate()` | No | No | Both positions and values are needed. |
| List comprehension | Yes | Yes | The transformation is short and clear. |

A **generator expression**, such as `sum(1 / x for x in data)`, can supply values
as needed without building a complete intermediate list. We will use this idea
when summing matrix rows. A generator is consumed as it is traversed; it is not a
reusable list.

In [20]:
# Assignment creates an alias; copy() creates a separate outer list.
original_values = [2, 4, 8]
alias = original_values
separate_copy = original_values.copy()

for index, value in enumerate(original_values):
    original_values[index] = 1 / value

print("Original list after replacement:", original_values)
print("Alias sees the same changes:", alias)
print("Separate copy:", separate_copy)

Original list after replacement: [0.5, 0.25, 0.125]
Alias sees the same changes: [0.5, 0.25, 0.125]
Separate copy: [2, 4, 8]


Here the list contains immutable numbers, so a shallow `.copy()` is sufficient
to preserve those values. Copying a list of lists requires more care: its inner
lists would still be shared. We will return to object references in a later lesson.

<a id="nested"></a>
## 5. Sequential and nested loops

**Sequential loops** run one after another. A **nested loop** is inside another
loop. For each item supplied to the outer loop, the inner loop runs before the
outer loop proceeds to its next item, unless control is interrupted.

More than two levels of nesting are possible. An inner loop's boundaries may
depend on the current outer-loop value. Let us first compare two simple cases.

### Sequential loops: pseudocode

```text
FOR i ← 1 TO 2, inclusive
    PRINT i
END FOR
FOR j ← 1 TO 3, inclusive
    PRINT j
END FOR
```

**Expected sequence:** `1, 2, 1, 2, 3`.

### Sequential loops: Python

`range(2)` provides two values, 0 and 1. Adding 1 when printing gives the
one-based display used in the pseudocode. The second loop starts only after the
first finishes, producing $2 + 3 = 5$ printed values.

In [21]:
for i in range(2):
    print(i + 1, end=' ')
for j in range(3):
    print(j + 1, end=' ')
print()

1 2 1 2 3 


### Nested loops: pseudocode

```text
FOR i ← 1 TO 2, inclusive
    FOR j ← 1 TO 3, inclusive
        PRINT (i, j)
    END FOR
END FOR
```

**Expected pairs:** `(1, 1), (1, 2), (1, 3), (2, 1), (2, 2), (2, 3)`.
The inner loop traverses all three values for each of the two outer values.

### Nested loops: Python

Here we use explicit starting values so the output matches the pseudocode
exactly. There are $2 \times 3 = 6$ executions of the inner body.

In [22]:
for i in range(1, 3):
    for j in range(1, 4):
        print((i, j))

(1, 1)
(1, 2)
(1, 3)
(2, 1)
(2, 2)
(2, 3)


### The position of a statement changes how often it runs

The outer-loop message below appears twice, while the inner-loop message appears
six times. Follow the indentation as you read the output.

In [23]:
for i in range(2):
    print(f"Outer iteration: {i + 1}")
    for j in range(3):
        print(f"    Inner iteration: {j + 1}")

Outer iteration: 1
    Inner iteration: 1
    Inner iteration: 2
    Inner iteration: 3
Outer iteration: 2
    Inner iteration: 1
    Inner iteration: 2
    Inner iteration: 3


### How does the amount of work grow?

The following counts assume that each body performs constant work and that there
is no early exit. Big-$O$ notation describes an upper bound on growth as the input
size increases, rather than an exact running time in seconds.

| Pattern | Body executions | Growth |
|---|---:|---|
| Sequential loops of lengths $n$ and $m$ | $n + m$ | $O(n + m)$ |
| Nested loops of lengths $n$ and $m$ | $nm$ | $O(nm)$ |
| Two nested loops, both of length $n$ | $n^2$ | $O(n^2)$ |
| A fixed $k$ nested levels, each of length $n$ | $n^k$ | $O(n^k)$ |

Nested loops do not automatically imply exponential complexity. With two full
loops of length $n$, doubling $n$ gives four times as many inner-body executions.
The later dependent-bound example requires a sum of inner-loop lengths.

<a id="matrices"></a>
## 6. Matrices and flattening
### Summing every element of a matrix

Consider the matrix

$$A = \begin{bmatrix}1 & 2 & 3 \\ 4 & 5 & 6\end{bmatrix}.$$

**Input:** matrix $A$. **Output:** the sum of all its entries, which is **21**.

```text
total ← 0
FOR each row index i
    FOR each column index j in row i
        total ← total + A[i, j]
    END FOR
END FOR
REPORT total
```

In this notebook, a matrix is represented as a list of row lists. Python accesses
an entry with `matrix[i][j]`; both positions start at zero.

### Index-based matrix traversal

Use row lengths instead of hard-coding the dimensions. This makes the loop
easier to reuse and also permits rows of different lengths, although such a
structure is no longer a rectangular mathematical matrix.

In [24]:
matrix = [[1, 2, 3], [4, 5, 6]]
matrix_total = 0
for i in range(len(matrix)):
    for j in range(len(matrix[i])):
        matrix_total += matrix[i][j]
print("Matrix total:", matrix_total)

Matrix total: 21


### Direct row and value traversal

When positions are unnecessary, iterate over the rows and then their values.
The next example uses a $3 \times 3$ matrix; its total is **45**. The accumulator
must actually be updated—printing each value alone would only inspect the data.

In [25]:
matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
matrix_total = 0
for row in matrix:
    print("Row:", row)
    for value in row:
        matrix_total += value
print("Matrix total:", matrix_total)

Row: [1, 2, 3]
Row: [4, 5, 6]
Row: [7, 8, 9]
Matrix total: 45


### Combining iteration with the built-in `sum()`

Summing each row and then summing the row totals gives the same result. A
generator expression avoids constructing a separate list of row totals.
Keeping variable names such as `total` also avoids hiding the built-in `sum()`.

In [26]:
matrix = [[1, 2, 3], [4, 5, 6]]
total_with_list = sum([sum(row) for row in matrix])
total_with_generator = sum(sum(row) for row in matrix)
print("With a list comprehension:", total_with_list)
print("With a generator expression:", total_with_generator)
# Later, NumPy will provide array-based approaches for numerical computation.

With a list comprehension: 21
With a generator expression: 21


### Flattening one level of nested data

Some data-processing tasks require a sequence of small groups to become a single
list. For example,

```text
[(1, 2), (3, 4), (5, 3), (9, 1), (4, 5)]
```

becomes

```text
[1, 2, 3, 4, 5, 3, 9, 1, 4, 5]
```

This operation preserves traversal order and flattens exactly one nesting
level. It is not a general recursive flattening procedure.

In [27]:
pairs = [(1, 2), (3, 4), (5, 3), (9, 1), (4, 5)]
flattened = []
for pair in pairs:
    for value in pair:
        flattened.append(value)
print(flattened)

[1, 2, 3, 4, 5, 3, 9, 1, 4, 5]


In the equivalent comprehension, read the `for` clauses **from left to right**:
first obtain a `pair`, then obtain each `value` from that pair. Their order
matches the explicit nested loop above.

In [28]:
pairs = [(1, 2), (3, 4), (5, 3), (9, 1), (4, 5)]
flattened = [value for pair in pairs for value in pair]
print(flattened)

[1, 2, 3, 4, 5, 3, 9, 1, 4, 5]


<a id="dependent"></a>
## 7. Nested loops with dependent bounds

An inner range can depend on the current outer-loop value. It is constructed
afresh each time execution reaches the inner loop. In the following example,
`j` starts at `i`, so only positions on or above the main diagonal are visited.
We add 1 when displaying positions to make them one-based.

In [29]:
size = 4
visited_count = 0
for i in range(size):
    for j in range(i, size):
        print((i + 1, j + 1), end=' ')
        visited_count += 1
print()
print("Number of visited positions:", visited_count)

(1, 1) (1, 2) (1, 3) (1, 4) (2, 2) (2, 3) (2, 4) (3, 3) (3, 4) (4, 4) 
Number of visited positions: 10


The inner loop runs 4, 3, 2, and 1 times, for a total of **10** positions. For a
square matrix of size $n$, including the diagonal gives

$$n + (n-1) + \cdots + 1 = \frac{n(n+1)}{2}.$$

Thus, the work still grows quadratically, $O(n^2)$, despite visiting fewer entries
than a full $n \times n$ traversal. To visit the lower triangle including the
diagonal, think about which column indices satisfy $j \leq i$; you will implement
that idea in Exercise 4.

<a id="control"></a>
## 8. Loop control: `break`, `continue`, and `else`

This section completes the loop-control topic in the original outline. It also
provides a brief preview of the next lesson on conditional logic.

`if condition:` runs its indented body only when the condition is true. The
operator `==` tests equality; `%` gives the remainder. For example,
`number % 2 == 0` tests whether an integer is even.

| Statement | Effect |
|---|---|
| `break` | Leave the nearest enclosing loop immediately. |
| `continue` | Skip the rest of the current iteration of the nearest loop. |
| `pass` | Do nothing; act as a syntactic placeholder. |

Official explanations appear in the
[Python control-flow tutorial](https://docs.python.org/3/tutorial/controlflow.html).

### Stop a search with `break`

Once the target has been found, the remaining items need not be inspected. Here,
the loop visits 4, 9, and 12, then stops before reaching 15 or 20.

In [30]:
values = [4, 9, 12, 15, 20]
target = 12
for value in values:
    print("Inspecting:", value)
    if value == target:
        print("Target found.")
        break

Inspecting: 4
Inspecting: 9
Inspecting: 12
Target found.


### Skip an item with `continue`

The reciprocal of zero is undefined. We can skip zeros and keep processing the
remaining values. Here, omitting zeros is an explicit teaching choice; a real
analysis would need a documented policy for such observations.

In [31]:
raw_values = [4, 0, 2, 0, 8]
valid_reciprocals = []
skipped_zeros = 0
for value in raw_values:
    if value == 0:
        skipped_zeros += 1
        continue
    valid_reciprocals.append(1 / value)

print("Reciprocals:", valid_reciprocals)
print("Skipped zeros:", skipped_zeros)

Reciprocals: [0.25, 0.5, 0.125]
Skipped zeros: 2


### Loop `else`: what happens when no `break` occurs?

A loop's `else` block runs after its iterable is exhausted normally, including
when the iterable is empty. It does not run when that loop ends through `break`.
This is useful for a search that needs a “not found” outcome. The `else` aligns
with `for`, not with the `if` inside it. An exception or `return` would also
prevent execution from reaching this `else` block.

In [32]:
values = [4, 9, 12, 15, 20]
for target in [12, 99]:
    for value in values:
        if value == target:
            print(f"{target}: found")
            break
    else:
        print(f"{target}: not found")

12: found
99: not found


### `break` affects one loop; `pass` does not skip anything

The first example exits each inner loop when `j` reaches 2, but the outer loop
still proceeds to its next value. In the second example, `pass` leaves execution
unchanged, so all three values are printed.

In Python, the pass statement is a placeholder that does absolutely nothing when executed. You use it in a for loop when Python requires a line of code to be there legally (syntactically), but you do not want any action to happen yet.

In [33]:
for i in range(2):
    for j in range(4):
        if j == 2:
            break
        print("Visited pair:", (i, j))

for number in range(3):
    if number == 1:
        pass
    print("After pass:", number)

Visited pair: (0, 0)
Visited pair: (0, 1)
Visited pair: (1, 0)
Visited pair: (1, 1)
After pass: 0
After pass: 1
After pass: 2


<a id="practice-checks"></a>
## 9. Common mistakes and reliable practice

| Mistake | Consequence | Better approach |
|---|---|---|
| Forgetting that `stop` is excluded | The last intended value is missing. | Inspect a short `list(range(...))`. |
| Using a step in the wrong direction | The range may be empty. | Check the start, stop, and sign together. |
| Using step 0 or a floating-point argument in `range()` | `ValueError` or `TypeError`, respectively. | Use integer counters and compute other values in the body. |
| Resetting an accumulator inside the loop | Previous work is discarded. | Initialize it before the loop. |
| Misplacing indentation or omitting `:` | Incorrect execution or a syntax error. | Identify which statements must repeat. |
| Assigning to the target to control iteration | The next item still comes from the iterator. | Choose the iterable or control statement deliberately. |
| Naming a variable `sum`, `list`, or `range` | A built-in name is hidden. | Use descriptive names such as `total` or `values`. |
| Appending or deleting during traversal of the same list | Items may be revisited or skipped. | Build a new list or traverse a deliberate copy. |
| Consuming the same generator twice | The second traversal has no remaining items. | Recreate it or store a reusable list. |

For a sum, initialize the accumulator to 0; for a product, use 1; for list
construction, use `[]`. Test both a small ordinary case and a relevant boundary
case, such as an empty input.

### An integrated example: summarizing valid scores

Let us combine traversal, selection, counting, and accumulation. For this
synthetic example, valid scores lie between 0 and 100 inclusive, and `None`
denotes a missing value. The validity condition uses `or`, which short-circuits:
when a value is `None`, Python does not evaluate the numeric comparison.

Try changing `raw_scores` to `[None, -5]` afterward. The second branch avoids
dividing by zero when there are no valid observations.

In [34]:
raw_scores = [80, None, 95, -5, 70, 110, 85]
valid_total = 0
valid_count = 0
rejected_count = 0

for score in raw_scores:
    if score is None or not (0 <= score <= 100):
        rejected_count += 1
        continue
    valid_total += score
    valid_count += 1

print("Valid observations:", valid_count)
print("Rejected observations:", rejected_count)
if valid_count > 0:
    mean_score = valid_total / valid_count
    print(f"Mean valid score: {mean_score:.2f}")
else:
    print("No valid observations; the mean is undefined.")

Valid observations: 4
Rejected observations: 3
Mean valid score: 82.50


**Check the interpretation:** four observations are valid and three are rejected.
The valid values sum to 330, so their mean is **82.50**. Rejected observations do
not enter either the numerator or the denominator. This example assumes that
every nonmissing item is numeric; validating arbitrary text inputs is beyond
the present lesson.

<a id="exercises"></a>
## 10. Exercises

Please attempt each task before opening the worked solutions. For prediction
tasks, write the entire output sequence and explain how indentation determines
it. Pseudocode boundaries in this section are inclusive.

### Exercise 1 · Trace a nested loop

What is the output of the following algorithm?

```text
FOR i ← 1 TO 3
    FOR j ← 1 TO 2
        PRINT i
    END FOR
END FOR
```

In [35]:
# Exercise 1 workspace: record your prediction, then implement the algorithm.

### Exercise 2 · Trace statements at different depths

What is the output? How many times does each `PRINT` statement execute?

```text
FOR i ← 1 TO 3
    PRINT i
    FOR j ← 1 TO 2
        PRINT j
    END FOR
END FOR
```

In [36]:
# Exercise 2 workspace: predict the output and explain the two print counts.

### Exercise 3 · Increasing counter, decreasing output

Write a loop with an **increasing** integer counter that produces

```text
5, 4, 3, 2, 1, 0
```

Use `range(6)` and transform the counter in the body. Do not use a negative step
for this exercise. A negative step would be valid Python, but the purpose here
is to distinguish the counter from the value being displayed.

In [37]:
# Exercise 3 workspace: use an increasing counter to create decreasing values.

### Exercise 4 · Sum a lower triangular region

Given

$$A = \begin{bmatrix}
1 & 0 & 3 & 7 \\
4 & 5 & 1 & 2 \\
1 & 2 & 1 & 4 \\
1 & 1 & 1 & 8
\end{bmatrix},$$

write Python code that sums the entries in the lower triangle, **including the
main diagonal**. In zero-based indexing, include positions satisfying $j \leq i$.

Use nested `for` loops and suitable bounds. Do not use `if`, `while`, a
user-defined function, or an aggregation helper such as `sum()`. The built-ins
`range()`, `len()`, and `print()` are allowed to construct loops and display the
answer. This makes the intended loop-only restriction explicit.

**Extension:** how would your inner range change to exclude the diagonal?

In [38]:
# Exercise 4 workspace.
exercise_matrix = [
    [1, 0, 3, 7],
    [4, 5, 1, 2],
    [1, 2, 1, 4],
    [1, 1, 1, 8],
]
# Add your nested loops here.

### Exercise 5 · Produce fractional values

Write a simple `for` loop that generates the numerical sequence

$$\frac{1}{2},\quad 0,\quad -\frac{1}{2},\quad -1.$$

Do not use `if` or `while`. Use an integer counter and calculate the required
value in the body; `range()` does not accept a floating-point step. Decimal
output such as `0.5` is acceptable.

In [39]:
# Exercise 5 workspace: generate the four values with an integer counter.

### Exercise 6 · Combine parallel data

A student has marks `[72, 80, 91]` for `['Quiz', 'Project', 'Exam']`, with weights
`[0.2, 0.3, 0.5]`, respectively. These weights sum to 1.

1. Use `zip()` to visit each assessment, mark, and weight together.
2. Display each weighted contribution.
3. Accumulate the weighted final mark, $\sum_i w_i x_i$.

**Extension:** use `enumerate()` around `zip()` to number the displayed rows.

In [40]:
# Exercise 6 workspace.
assessments = ['Quiz', 'Project', 'Exam']
marks = [72, 80, 91]
weights = [0.2, 0.3, 0.5]
# Add your loop and accumulator here.

### Exercise 7 · Search and report failure

Search `[3, 7, 7, 10]` for a target. Report its first zero-based position and stop
immediately. Use a loop `else` to report when the target is absent. Try both
`target = 7` and `target = 9`.

**Reflection:** why would a `continue` statement fail to stop the search at the
first match?

In [41]:
# Exercise 7 workspace.
search_values = [3, 7, 7, 10]
target = 7
# Add your search here; then try target = 9.

<a id="review"></a>
## 11. Review and next steps

We have used the same basic mechanism—visiting successive items—to sum integers,
transform observations, traverse matrices, and search for a value.

Before moving on, check that you can explain:

- Why `range(1, 101)` includes 100, and why some ranges are empty.
- Why an accumulator usually needs initialization before its loop.
- When direct iteration, indexing, `enumerate()`, or `zip()` is appropriate.
- How a new output list differs from replacing elements in an existing list.
- Why nested-loop counts depend on the lengths of their inner traversals.
- How `break`, `continue`, `pass`, and loop `else` differ.

**Next lesson:** conditional logic with `if`, including its combination with
loops. Later lessons will revisit reusable functions, object references, and
numerical array operations with NumPy.

I suggest one final check: choose an example, explain it aloud without running
it, then restart the kernel and run the notebook from the beginning.

**End of Week II.** The optional worked solutions follow.

<a id="solutions"></a>
## Appendix · Worked solutions

These are reference implementations, not the only acceptable answers. Compare
your reasoning as well as your output. The solution cells use separate variable
names so they do not depend on unfinished exercise workspaces.

### Solutions 1 and 2 · Trace the output

Exercise 1 prints each outer value twice, producing `1, 1, 2, 2, 3, 3`.
Exercise 2 produces `1, 1, 2, 2, 1, 2, 3, 1, 2`: the outer `PRINT` runs three
times, and the inner `PRINT` runs six times.

In [42]:
solution_1 = []
for i in range(1, 4):
    for j in range(1, 3):
        solution_1.append(i)
print("Exercise 1:", solution_1)

solution_2 = []
for i in range(1, 4):
    solution_2.append(i)
    for j in range(1, 3):
        solution_2.append(j)
print("Exercise 2:", solution_2)

Exercise 1: [1, 1, 2, 2, 3, 3]
Exercise 2: [1, 1, 2, 2, 1, 2, 3, 1, 2]


### Solution 3 · Transform the counter

As `i` increases from 0 to 5, the expression `5 - i` decreases from 5 to 0.

In [43]:
solution_3 = []
for i in range(6):
    solution_3.append(5 - i)
print(solution_3)

[5, 4, 3, 2, 1, 0]


### Solution 4 · Include the diagonal

For row `i`, `range(i + 1)` supplies columns 0 through `i`. The included row sums
are 1, 9, 4, and 11, giving a total of **25**. To exclude the diagonal, use
`range(i)` instead; the total would be **10**.

In [44]:
solution_matrix = [
    [1, 0, 3, 7],
    [4, 5, 1, 2],
    [1, 2, 1, 4],
    [1, 1, 1, 8],
]
lower_total = 0
for i in range(len(solution_matrix)):
    for j in range(i + 1):
        lower_total += solution_matrix[i][j]
print("Lower triangle including the diagonal:", lower_total)

strict_lower_total = 0
for i in range(len(solution_matrix)):
    for j in range(i):
        strict_lower_total += solution_matrix[i][j]
print("Lower triangle excluding the diagonal:", strict_lower_total)

Lower triangle including the diagonal: 25
Lower triangle excluding the diagonal: 10


### Solution 5 · Compute fractional values from integers

The expression $(1-i)/2$ gives the requested values for $i = 0, 1, 2, 3$.

In [45]:
solution_5 = []
for i in range(4):
    solution_5.append((1 - i) / 2)
print(solution_5)

[0.5, 0.0, -0.5, -1.0]


### Solution 6 · Accumulate weighted contributions

The weighted contributions are 14.4, 24.0, and 45.5, giving a final mark of
**83.9**. If weights did not sum to 1, a weighted mean would also require division
by their sum. This example includes the optional row numbering.

In [46]:
solution_assessments = ['Quiz', 'Project', 'Exam']
solution_marks = [72, 80, 91]
solution_weights = [0.2, 0.3, 0.5]
weighted_total = 0

for position, (name, mark, weight) in enumerate(
    zip(solution_assessments, solution_marks, solution_weights, strict=True), start=1
):
    contribution = mark * weight
    weighted_total += contribution
    print(f"{position}. {name}: {mark} × {weight:.1f} = {contribution:.1f}")
print(f"Weighted final mark: {weighted_total:.1f}")

1. Quiz: 72 × 0.2 = 14.4
2. Project: 80 × 0.3 = 24.0
3. Exam: 91 × 0.5 = 45.5
Weighted final mark: 83.9


### Solution 7 · Stop at the first match

The first occurrence of 7 is at position 1; 9 is absent. `break` stops the inner
search loop, whereas `continue` would only move it to the next iteration.

In [47]:
solution_values = [3, 7, 7, 10]
for target in [7, 9]:
    for position, value in enumerate(solution_values):
        if value == target:
            print(f"Target {target}: first found at position {position}")
            break
    else:
        print(f"Target {target}: not found")

Target 7: first found at position 1
Target 9: not found


### Checks: compare computed results with known answers

`assert` checks a condition and raises an error if it is false. For floating-point
arithmetic, `math.isclose()` is useful when tiny representation differences may
occur. These checks are a compact demonstration of verifying our reasoning.
They refer to the worked solutions immediately above.

In [48]:
from math import isclose

assert solution_1 == [1, 1, 2, 2, 3, 3]
assert solution_2 == [1, 1, 2, 2, 1, 2, 3, 1, 2]
assert solution_3 == [5, 4, 3, 2, 1, 0]
assert lower_total == 25 and strict_lower_total == 10
assert solution_5 == [0.5, 0.0, -0.5, -1.0]
assert isclose(weighted_total, 83.9)
print("All worked-solution checks passed.")

All worked-solution checks passed.


<a id="references"></a>
## References and further reading

Consult the official Python documentation:

1. [The `for` statement and `range()`](https://docs.python.org/3/tutorial/controlflow.html#for-statements)
   — introductory explanations and examples.
2. [`break`, `continue`, and loop `else`](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements)
   — controlling and completing iteration.
3. [List comprehensions and looping techniques](https://docs.python.org/3/tutorial/datastructures.html)
   — building lists and traversing collections.
4. [The `range` sequence type](https://docs.python.org/3/library/stdtypes.html#ranges)
   — boundaries, steps, and representation.
5. [Built-in functions](https://docs.python.org/3/library/functions.html)
   — `enumerate()`, `zip()`, and `sum()`.